### PR TITLE
PR: Have LSP update after closequotes and closebrackets

### LINKED ISSUE #9323: Close brackets or close quotes doesn't update the LSP
### Issue Report Checklist

* [X] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [X] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [X] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [ ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description

Adding brackets or quotes with close_brackets or close_quotes enabled doesn't update the code error checking
`print("sweet)`

### What steps reproduce the problem?

1. `print("sweet)` gives error EOL sting literal error
2. type a quote to make `print("sweet")`
3. The error stays

### What is the expected output? What do you see instead?

No errors!

### Paste Traceback/Error Below (if applicable)

```python-traceback

PASTE TRACEBACK HERE

```

## Versions

* Spyder version: master
* Python version: 3.6
* Qt version: 
* PyQt version: 
* Operating System name/version: Linux

### Dependencies

```
PASTE DEPENDENCIES HERE
```

### PR DESCRIPTION
**Pull request by bcolsen**
## Description of Changes

Add document_did_change to closequotes and closebrackets so the LSP will update.

* [ ] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

The LSP would remove errors after typing brackets or quotes to correct them. Now it does.

### Issue(s) Resolved

Fixes #9323

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: bcolsen

### CODE REVIEW

**Comment by bcolsen** [comment]
I got thinking about this when I saw #9300. The tests for that PR were put in test_warnings is there a similar test_error file? Quote and brackets tend to make errors not warnings.

We might want this in beta2

**Comment by goanpeca** [comment]
@bcolsen could we add a test for this?

Maybe add something here https://github.com/spyder-ide/spyder/blob/master/spyder/plugins/editor/widgets/tests/test_warnings.py ?

**Comment by bcolsen** [comment]
@goanpeca That would work but I can't think of a warning that would involve quotes or brackets. You usually get errors in that case.

**Comment by goanpeca** [comment]
I don't follow... those tests (the file I attached) include warnigns and erros etc...

**Comment by bcolsen** [comment]
> I don't follow... those tests (the file I attached) include warnings and errors etc... (not only "warnings")

Sorry for the confusion. I thought there was some differentiation between warnings and errors in the code, but `editor.get_current_warnings()` gets both so that works.

**Comment by ccordoba12** [comment]
Thanks a lot @bcolsen! Please fix the tests you added so we can include this one in beta2.

**Comment by bcolsen** [comment]
I can't get the test to run locally on my computer. It seem like the LSP service doesn't start because it hits the time out error. Can you think of anything I might be missing?

**Comment by ccordoba12** [comment]
I restricted the lastet version of the PyLS we support to be less than 0.25 in the last couple of days. So that could be what's not allowing you to run our tests locally.

File: spyder/plugins/editor/widgets/tests/test_warnings.py [hunk 1]
```diff
@@ -175,3 +177,71 @@ def test_update_warnings_after_delete_line(qtbot, lsp_codeeditor):
...
+    assert editor.get_current_warnings() == expected
+
+    # Add a single quote to fix the error
+    editor.move_cursor(-2)
+    qtbot.keyPress(editor, Qt.Key_Apostrophe)
+    assert editor.toPlainText() == "print('test')\n"
+
+    # Wait for the lsp_response_signal.
+    with qtbot.waitSignal(editor.lsp_response_signal, timeout=30000):
+        sleep(1)
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_warnings.py]: Why is this necessary? I think instead of this you need to run `editor.document_did_change()`.

Comment (bcolsen) [reply spyder/plugins/editor/widgets/tests/test_warnings.py]: I'm not entirely  sure honestly but in my mind I'm  waiting for the server to be done processing.

> I think instead of this you need to run editor.document_did_change()

The only change I made to the the spyder code other than tests was adding editor.document_did_change() to close quotes and brackets. I don't  know how editor.document_did_change() works but if I just call editor.document_did_change() here wouldn't that defeat the purpose of the test?

Comment (goanpeca) [reply spyder/plugins/editor/widgets/tests/test_warnings.py]: I agree with @bcolsen. I think if we are using `editor.document_did_change()` on the tests, we are probably doing something wrong as the editor should be doing this "on its own". 

Now the way to use the context blockers should be:

```python
with qtbot.waitSignal(editor.lsp_response_signal, timeout=30000):
    # do something (add a quote.... remove a quote) no need to add waits
```

I suggest you put the code that triggers something inside the `with` instead of document change or a wait. The code inside the with should be triggering some signal.

Comment (bcolsen) [reply spyder/plugins/editor/widgets/tests/test_warnings.py]: I tried that first, but I couldn't get it to fail and I needed to add the sleep in side as well to make it pass.

Comment (ccordoba12) [reply spyder/plugins/editor/widgets/tests/test_warnings.py]: @bcolsen, ok, you are the one who fought with the test, so I defer to you on this one. Please add a comment above this line then, explaining why it's needed.

Also, why didn't you use `qtbot.wait` instead?

Comment (bcolsen) [reply spyder/plugins/editor/widgets/tests/test_warnings.py]: I didn't try qtbot.wait. That should work.

Comment (ccordoba12) [reply spyder/plugins/editor/widgets/tests/test_warnings.py]: Yeah, I think it's better because it's non-blocking (while `sleep` is).

Comment (goanpeca) [reply spyder/plugins/editor/widgets/tests/test_warnings.py]: Yep! correct. 

@bcolsen I will check later what might be the problem that requires manually calling `document_did_change` Thanks for the work here!

**Comment by bcolsen** [comment]
I tried the tests on my work computer today and it works with the code in the `with` block now. That makes the test cleaner.  However, when I remove the `document_did_change` from the extensions the test just stalls and times out at the `with` statement instead of returning the error.

I think this might might indicate something deeper and that manually calling `document_did_change` in the extensions might just be a workaround, but does work :-)

File: spyder/plugins/editor/widgets/tests/test_warnings.py [hunk 2]
```diff
@@ -8,9 +8,11 @@
 
 # Stdlib imports
 import os
+from time import sleep
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/tests/test_warnings.py]: Please remove this line because it's not needed anymore. The rest looks good.

**Comment by goanpeca** [comment]
Great work! :-)

**Comment by bcolsen** [comment]
Thank you. That was a rough "one line change"  ;-)
